[K] 미션 4 모든 기물 배치하기

kihyuk-sung opened the pull request:
utils.StringUtils 클래스 추가
- String에 줄바꿈을 추가하는 appendNewLine() 메소드
- 줄바꿈 문자열인 NEWLINE 필드

Piece 클래스 추가
- Pawn 클래스에서 이름을 변경하고 기능을 추가하였습니다.
- Piece는 색상으로 Color Enum, 이름으로 Type Enum을 가집니다.
- Type은 PAWN, KNIGHT, ROOK, BISHOP, QUEEN, KING 입니다.
- getColor()와 getRepresentation() 메소드를 가집니다.
- getRepresentation()은 Color와 Type에 따라 정해집니다.
- 색과 말의 종류 마다 static 팩토리 메소드를 가집니다. (createWhitePawn(), createBlackQueen()...)

Board 클래스 수정
- 이제 Piece 객체를 가지고 초기화합니다.
- public으로 initialize(), pieceCount(), showBoard() 메소드가 있습니다.
- 내부에 Piece를 관리하기 위한 List가 Map<Color, List<Pawn(Piece)>> 였던 것을 4개의 List<Piece>로 수정하였습니다.
- Piece의 생성이 이제 Color Enum을 패러미터로 받지 않고 static 팩토리 메소드로 생성하여, Color를 이용하여 Board의 다른 메소드들을 합치기 어려워졌기 때문입니다.

ChessGame main 메소드 리팩토링
- 기능별로 메소드를 추가하여 가독성을 높이기 위해 노력했습니다.

전체적으로 else를 제거하려고 노력하였습니다.

그럼 리뷰 잘부탁드립니다!

kihyuk-sung committed:
test: Add StringUtilsTest

- StringUtils has appendNewLine(String string) method. This method will
add line seperator to string and return it.
- StringUtils has NEWLINE field which is line seperator.

- StringUtils는 appendNewLine(String string) 메소드를 가집니다. 이
메소드는 줄바꿈을 string에 추가하여 반환합니다.
- StringUtils는 줄바꿈 문자열인 NEWLINE field를 가집니다.

kihyuk-sung committed:
feat: StringUtils class

- StringUtils has appendNewLine(String string) method. it will add line
seperator to string and return it.
- StringUtils has NEWLINE string field. it equals to
System.lineseperator().
- Because Fields and methods in StringUtils are static, its constructor
is private.

- StringUtils에 appendNewLine(String string) 메소드가 있습니다. string에
줄바꿈을 추가하여 반환합니다.
- NEWLINE field도 있습니다. 이는 System.lineseperator()와 같습니다.
- StringUtils의 필드와 메소드는 static이기 때문에, 생성자를 private로
하였습니다.

kihyuk-sung committed:
test: Edit BoardTest code

- To test adding New Pawn Object.

kihyuk-sung committed:
refactor: Change switch-case-default to if in Pawn

- For code readability.

kihyuk-sung committed:
refactor: Add FILES field, refactor print() method

- Add FILES to maintain board column.
- refactor print() to reduce code duplication.

kihyuk-sung committed:
refactor: Board.print() using utils.StringUtils

- for more readability.

kihyuk-sung committed:
teat: Add PieceTest to test Piece creation

- Pawn is renamed to Piece.
- Piece has static Factory method.

kihyuk-sung committed:
feat: Piece class

- Piece has Color and Type.
- Color is White and Black. So, use Color Enum.
- Type is Pawn, Knight, Bishop, Rook, Queen, King. So, use Type Enum.
- Piece has getColor(), getRepresentation() method.
- getColor() returns a Color of Piece.
- getRepresentation returns a letter of Piece.
- Black Piece is Upper Case letter and White Piece is Lower Case letter.
- Piece has static factory methods such as create{Color}{Type}.

- Piece는 Color와 Type을 가집니다.
- Color는 흰색과 검은색을 가지므로, Color enum을 사용합니다.
- Type은 Pawn, Knight, Bishop, Rook, Queen, King 입니다. 그래서 Type Enum
을 이용하였습니다.
- Piece는 getColor(), getRepresentation() 메소드를 가집니다.
- getColor() 는 Piece의 Color를 반환합니다.
- getRepresentation()은 Piece의 문자를 반환합니다.
- 검은색 Piece는 대문자, 흰색 Piece는 소문자입니다.
- Piece는 각 색상과 타입의 인스턴스를 반환하는 static 팩토리 메소드를
가집니다.

kihyuk-sung committed:
test: Edit BoardTest

- Board will use Piece and initialize all pieces.
- Board will have pieceCount(), showBoard() methods.

- Board는 Piece를 이용할 것이고, 모든 Pieces를 초기화 합니다.
- pieceCount(), showBoard() 메소드를 가집니다.

kihyuk-sung committed:
feat: Initialize all pieces in Board class

- 이제 Board는 Piece 클래스를 사용합니다.
- Pawn 클래스는 삭제하였습니다.
- public 메소드인 pieceCount(), showBoard(), initialize()를 제외하고
나머지 메소드는 private로 바꾸었습니다.
- pawnList를 다시 색별로 나누었습니다. 외부에 제공했던 getPawnResult(Color)
를 없앴고, Piece 또한 외부 인터페이스로 Color를 사용하지 않도록 변경되었기
때문에, Board에서도 Color를 사용하지 않는 것이 좋겠다 생각했습니다.
- 그래서 Pawn 외의 다른 Piece들을 관리하는 List도 색깔 별로 추가하였습니다.

- size()는 pieceCount()로 변경하였습니다.
- print()를 showBoard()로 변경하였습니다.

kihyuk-sung committed:
test: Piece.isBlack() Piece.isWhite()

- This methods are checking the color of Piece.

kihyuk-sung committed:
feat: Piece.isBlack() and Piece.isWhite()

- check the color of Piece is black or white.

kihyuk-sung committed:
refactor: main method

- To compact main method.

kihyuk-sung committed:
refactor: Remove else, and duplicate code

- For readablity.

kihyuk-sung committed:
refactor: Rename piece list and methods in Board

- 문법에 맞게 otherBlackPieceList, otherWhitePieceList로 수정
- getPiecesResult(): 복수로 변경

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,60 +1,84 @@
+    private final List<Piece> blackPawnList = new ArrayList<>();
+    private final List<Piece> otherBlackPieceList = new ArrayList<>();
+    private final List<Piece> whitePawnList = new ArrayList<>();
+    private final List<Piece> otherWhitePieceList = new ArrayList<>();
```
`Piece` 에서 `Color` 와 `Type` 을 관리하는데, 네 개의 `List` 객체를 두는 것은 좋은 선택이 아닌 것 같습니다.
하나의 `List` 로 관리하는 방법은 없을까요?

kihyuk-sung replied:
하나의 `List`로 관리하는 방법이라면 `List`를 합치고, 그 인덱스와 출력할 위치를 1대1 대응하는 방법이 제일 좋은 방법이라고 생각합니다. 
그런데 하나로 합쳐보니 initialize() 메소드를 실행하지 않고서 showBoard()를 실행하면 IndexOutOfBoundException()이 발생할 것입니다. 그래서 아예 고정 사이즈인 배열을 이용하였습니다. 다만, null을 직접적으로 이용하게 되었는데, 이 경우는 다음 미션에서 빈칸을 나타내는 Piece를 추가하면 문제가 없을 것입니다.

wheejuni replied:
`initialize()` 없이 `showBoard()` 가 실행되면 예외가 던져지는게 자연스러운 플로우는 아닐까요?

kihyuk-sung replied:
비어있는 상태의 게임 보드도 보여줄 수 있어야한다고 생각합니다.
initialize() 메소드는 보드 위에 말을 배치하는 것으로 생각했어요.

wheejuni replied:
`initialize()` 에서 빈 칸을 추가하는 작업도 수행해주면 될 것 같아요.

kihyuk-sung replied:
빈칸을 실제 클래스로 구현하는 것은 미션 5의 요구사항이라 이번 미션에선 구현하지 않았습니다.

wheejuni commented on src/main/java/chess/ChessGame.java:
```diff
@@ -3,41 +3,56 @@
+        isStarted = true;
+    }
+
+    public static void showStartMessage() {
+        System.out.println("체스 게임 시작");
+        System.out.println("게임 시작은 start, 종료는 end를 입력하세요");
+    }
```
콘솔에 직접 메세지를 출력하는 `void` 타입의 `public` 메소드를 두는 것은 _유닛테스트 하기 좋은 코드를 작성한다_ 는 관점에서 어떻게 평가할 수 있을까요? 콘솔에 출력된 메세지를 테스트하려면 어떻게 해야 할까요?

kihyuk-sung replied:
`public void` 메소드는 유닛테스트 하기엔 좋은 코드가 아닐 수 있습니다. 일반적으로 `public void` 인 메소드라고 하면 객체의 상태를 바꾸는 메소드가 있을 것입니다. 이런 메소드의 동작을 실제로 테스트하려면 그 변화를 얻을 수 있는 메소드를 추가해야할 것입니다.
그리고 콘솔에 고정적인 메시지를 출력하는 메소드는 너무 간단한 기능이라 생각하여 굳이 테스트를 할 생각을 하지 못했습니다. 실제로 좀 더 복잡해지면, 객체의 상태에 따라 메소드가 출력하는 메시지가 달라질 것입니다. 이를 테스트 하려면 반환타입을 String으로 바꾸고 반환한 값을 외부에서 표준출력 해주는 것이 제일 좋은 방법이라고 생각합니다.

[Unit Testing of System.out.println() with JUnit](https://www.baeldung.com/java-testing-system-out-println)

[JUnit 5 User-Guide Running Tests-Capturing Output](https://junit.org/junit5/docs/current/user-guide/#running-tests-capturing-output)

한편, 직접적으로 콘솔 출력을 테스트할 방법도 찾아보았습니다. 위 링크는 표준 출력을 리다이렉션하는 방법으로 출력을 가져오는 방법으로 실제 코드에 적용해보았습니다.

아래 링크는 JUnit 5 Platform에서 표준 출력/에러를 잡아내는 방법에 대해 적혀있는데,, TestExecutionListener를 구현하여 잡아내는 방법인듯 합니다. 다만 아직 JUnit5에 대한 지식이 부족하기도 하여 실제로 적용해보진 못했습니다.

wheejuni replied:
`너무 간단한 기능이라 생각하여 굳이 테스트를 할 생각을 하지 못했` 는데 별도의 메소드로 분리하실 필요가 있었을까요.
사실 이 질문은 테스트에 대한 얘기가 아니었습니다.
메소드의 분리 기준에 대한 의견이었는데요. 한번 다시 생각해 보시기를 권합니다.

kihyuk-sung replied:
메소드의 분리 기준은 기능을 하나 수행하면 그 묶음으로 분리할 수 있다고 생각합니다. 그래서 시작 전 메시지를 출력하는 것을 showStartMessage(), 비슷하게 명령어 리스트 출력으로 help()로 분리하였습니다. 이정도는 기능이 작기 때문에 메소드로 분리하지 않아도 괜찮다는 의견이신가요?

wheejuni reviewed (changes requested):
4단계까지 빠르게 진행해 주셨네요! 수고하셨습니다. 💯 
짧은 리뷰 의견입니다만, 생각해보실 거리는 많을 것 같습니다. 
특히 `List` 에 관한 리뷰 의견 꼭 읽어보시고 설계를 바꿀 수 있다면 다른 설계로 구현해보시면 좋겠습니다. 감사합니다!

kihyuk-sung committed:
test: ChessGame help(), showStartMessage()

- ChessGame의 표준출력만 하는 help(), showStartMessage() 메소드 테스트코드
- 표준 출력 리다이렉션으로 값을 가져옴.

kihyuk-sung committed:
refactor: Merge inner pieceLists to one list

- 내부에 있던 4개의 리스트를 하나로 합쳤습니다.
- index가 0~7은 첫번째줄, 8~15는 두번째줄, 16~23은 3번째줄, 24~31은 4번째줄
입니다.
- initialize()를 반드시 하고 showBoard()를 실행해야 합니다.

kihyuk-sung committed:
refactor: Change pieceList to Array

- List<Piece> 였을 때는 pieceList에 Piece 32개가 무조건 들어가있어야
showBoard()를 실행할 수 있었습니다.
- 그래서 고정 사이즈의 Piece Array를 사용하여 보드의 위치와 Array index가
일대일 대응이 되도록 하였습니다.

kihyuk-sung commented:
리뷰 감사드립니다! 리뷰 해주신 것에 대해 공부해보면서 어느 것이 나은지 고민을 하다보니 늦은 것 같습니다.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -2,83 +2,92 @@
-    private final List<Piece> otherBlackPieceList = new ArrayList<>();
-    private final List<Piece> whitePawnList = new ArrayList<>();
-    private final List<Piece> otherWhitePieceList = new ArrayList<>();
+    private final Piece[] pieceList;
```
배열을 사용하시기를 제안한 부분이 전혀 아니었습니다. ㅠㅠ

kihyuk-sung replied:
이전에 4개의 List로 구현한 것은 결국 List하나가 하나의 Row(Chess의 Rank)를 담당하기 때문이었습니다. 기존에 색깔 별로 존재하는 두개의 Pawn List를 확장한다는 측면에서도 4개의 List로 구현하는 것은 매우 자연스러운 것이라 생각합니다.

이를 하나의 List로 합친다면, 하나의 리스트가 하나의 Row를 출력할 것이라는 연관관계도 무너지게 될 것입니다. 그렇다면 리스트와 보드와의 추가적인 연관관계를 결정하지 않은 상황에서 Piece 객체의 Color와 Type 만으로 보드 위의 위치를 정하는 것은 불가능합니다. Piece 객체가 보드 위의 위치를 갖지 않는 한 말입니다. Piece 객체가 보드 위의 위치를 갖는다면 List의 크기를 한정하지 않고 Piece에게 Board 위의 위치를 정하는 것 까지 위임할 수 있을 것입니다.

다만 Piece 객체가 위치를 갖는 설계는 이후에 보드에 임의 접근하는 측면에서 매우 좋지 않은 설계가 될 것입니다. 매번 각 피스가 어느 위치에 존재하는지 찾아야 할테니까요. 또한 Piece의 위치는 게임을 진행하면서 매번 변할텐데, Piece 객체는 Value 객체라는 미션의 요구사항과도 맞지 않습니다.

그래서 하나의 List에 Piece 객체를 관리한다면 결국 그 List의 index와 보드의 위치와 1:1대응을 시키는 것이 맞을 것입니다. `initialize() 에서 빈 칸을 추가하는 작업도 수행해주면 될 것 같아요.` 라는 리뷰에서 리뷰어분도 또한 List 내의 Piece 객체의 위치와 보드와 연관관계가 있다는 것을 가정하에 이야기를 하셨을 것 같습니다.

List를 하나로 합치며, 그 인덱스와 실제 보드 위의 위치를 대응하는 구조를 생각해본다면, List의 구현체는 ArrayList가 제일 적합할 것입니다. 체스게임을 하면서 보드에 임의접근을 제일 많이 할 것이니까요.

여기서 저는 ArrayList 크기가 실제로 변할 일이 없다면 고정크기의 배열로 선언하는 것이 더 낫다고 생각했습니다. 적어도 게임 한판을 하는 동안에 체스 보드의 크기가 Rank 8 * File 8 이라는 것은 변하지 않을 것이니까요.

다만 이는 범용적이지 못한 설계인 것은 맞습니다. List 인터페이스는 여러 List에서 공유하기 때문에 누구나 알고 있고 배열 접근보다 읽기 편할 수 있으니까요. 한번 리스트로 구현했던 [c092465](https://github.com/codesquad-members-2021/java-chess/pull/90/commits/c09246530d56f85ea2672c3b469dc52ec415f3bb) 커밋에서 [7365281](https://github.com/codesquad-members-2021/java-chess/pull/90/commits/7365281f28e8e2cc0fbd4cf5c4a40cb6071d3c09) 커밋으로 넘어오면서 리스트와 관련하여 많은 수정이 생긴 것도 문제라고 할 수 있습니다. 그리고 제가 작성한 코드는 null을 직접 관리하고 있습니다. 현재 구현한 것은 빈칸을 제외한 32개의 배열을 선언하고, 혹시 모를 가능성을 위해 null에 대한 처리를 해준 것이기 때문입니다. 다만 이것은 이후에 미션 5에서 빈칸에 대응하는 Piece를 구현하면 개선할 수 있을 것입니다.

또한 배열로 구현한 것도 List로 똑같은 동작을 하도록 구현할 수 있었을 것입니다. 처음에 null을 채운 후, List.set()으로 값을 바꾸는 형태로요.

wheejuni reviewed (changes requested):
merge하기 이전에 두 개의 궁금증이 있습니다. 
질문 코멘트를 남겨드렸는데, K의 생각을 듣고 싶어요.

kihyuk-sung commented:
코멘트 확인하고 제 생각을 정리하여 추가로 달았습니다. 리뷰 감사드려요!

wheejuni commented:
@PizzaCola-K 타당한 생각을 갖고 코드 작성하신 것 같습니다. 추가 수정요청 없이 바로 merge 할게요.
